rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,112 @@
+
+## Enlarge the reserved range of system users/groups
+
+Since the new recommendation is to use `isSystemUser` for users and to not set static ids, the available ids now have the range `SYS_UID_MIN` to `SYS_UID_MAX` and `SYS_GID_MIN` to `SYS_GID_MAX`, defined in [shadow.nix](https://github.com/NixOS/nixpkgs/blob/044cc701c23ede96355eb1dc997985b4dfac0372/nixos/modules/programs/shadow.nix#L13-L21), both of which currently span from 400 to 499. This means a NixOS system could only have 100 different services using dynamically allocated ids over its lifetime. Since this is not very much, this range will be changed to span from 400 to 999, reserving an additional 500 ids for system users/groups. The range from 500 to 999 is currently not reserved for anything.
+
+Therefore the number of different services using dynamic ids that can be enabled on a single NixOS system is 600. In comparison with static ids, the number of *enabled* services on a system grows much slower than the number of total *existing* services in NixOS (each of which would need one of the 400 static ids).
+
+The implementation of this is in PR [#65698](https://github.com/NixOS/nixpkgs/pull/65698).
+
+A note on uids vs gids: If `isSystemUser = false`, dynamic uids are allocated in the range `UID_MIN` to `UID_MAX`, while with `isSystemUser = true`, it's `SYS_UID_MIN` to `SYS_UID_MAX`. In contrast, gids are always allocated in the range `SYS_GID_MIN` to `SYS_GID_MAX`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If data is restored from a backup without restoring `/var/lib/nixos` as well and the service doesn't [ensure correct directory permissions](#ensuring-correct-directory-permissions-for-services), then the service can fail to start. When copying data from a different machine this can happen as well. This is fixable by either manually changing `/var/lib/nixos/{u,g}id-map` to map the name to the old id, by recursively `chown`ing the restored data to the new id, or by making the NixOS service fix permissions itself.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Increase the range of static ids up to 999, adding another 600 static ids. While this would solve the problem of running out of static ids temporarily, over time this range will fill up again, at which point this alternative can't be used again.
+- Use a big range above `UID_MAX`/`GID_MAX`=29999, perhaps from 30000 to 40000. This would be enough static ids for a long time, but having to maintain a static mapping in nixpkgs will still be annoying.
+
+Not doing anything is not an option as the currently used range is finite.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+What is the migration path from different user declarations?
+
+| From \ To | `DynamicUser` | dynamic NixOS ids | static NixOS ids |
+| --- | --- | --- | --- |
+| `DynamicUser` | - | ??? | ??? |
+| dynamic NixOS ids | ??? | - | ??? |
+| static NixOS ids | ??? | ??? | - |
```

Comment:
I'm not sure this RFC is about any other case than "static NixOS ids" -> `DynamicUser`, and the migration path for that is the one you describe below in "Future work".

Or do I misunderstand what the unresolved question is?

Reply:
I think we should think (and test) how all of these behave, making sure they actually work and don't break services or how to ensure they don't:

- static NixOS id -> dynamic NixOS id: For migrating from static ids
- static NixOS id -> `DynamicUser`: For migrating from static ids
- `DynamicUser` -> dynamic NixOS id: When `DynamicUser` was found to be too restrictive
- dynamic NixOS id -> `DynamicUser`: When `DynamicUser` was found out to be feasible